File: projects/state-machine/python/src/history.py
"""
History management for state transitions.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Deque, Dict, List, Optional, Sequence

from .event import Event
from .state import State


@dataclass
class HistoryEntry:
    """
    Represents a single state transition in the history.

    Attributes:
        timestamp: When the transition occurred
        from_state: The source state
        to_state: The target state
        event: The triggering event
        success: Whether the transition was successful
        duration: Time spent in the previous state (in seconds)
        context: Optional context data
        error: Optional error if transition failed
    """

    timestamp: datetime
    from_state: State
    to_state: State
    event: Event
    success: bool
    duration: Optional[float] = None
    context: Optional[Dict[str, Any]] = None
    error: Optional[Exception] = None

    def __repr__(self) -> str:
        status = "OK" if self.success else "FAILED"
        return (
            f"HistoryEntry({self.timestamp.isoformat()}, "
            f"{self.from_state} -> {self.to_state} on {self.event}, {status})"
        )

    def __str__(self) -> str:
        status = "✓" if self.success else "✗"
        return (
            f"[{self.timestamp.strftime('%H:%M:%S')}] "
            f"{self.from_state} -> {self.to_state} on {self.event} {status}"
        )


class HistoryState:
    """
    Tracks the last active sub-state for hierarchical states.

    This allows hierarchical states to remember which sub-state
    was active when they were last exited.

    Example:
        >>> history = HistoryState()
        >>> history.save("Active", "Processing")
        >>> history.get("Active")
        'Processing'
    """

    def __init__(self) -> None:
        """Initialize the history state tracker."""
        self._states: Dict[str, str] = {}

    def save(self, parent_state: str, sub_state: str) -> None:
        """
        Save the last active sub-state for a parent state.

        Args:
            parent_state: The parent state name
            sub_state: The last active sub-state name
        """
        self._states[parent_state] = sub_state

    def get(self, parent_state: str) -> Optional[str]:
        """
        Get the last active sub-state for a parent state.

        Args:
            parent_state: The parent state name

        Returns:
            The last active sub-state name, or None
        """
        return self._states.get(parent_state)

    def clear(self, parent_state: Optional[str] = None) -> None:
        """
        Clear history for a specific or all parent states.

        Args:
            parent_state: Optional specific parent state to clear
        """
        if parent_state:
            self._states.pop(parent_state, None)
        else:
            self._states.clear()

    def __repr__(self) -> str:
        return f"HistoryState({self._states})"


class HistoryManager:
    """
    Manages the history of state transitions.

    Provides methods to query and analyze transition history.

    Example:
        >>> manager = HistoryManager(max_entries=100)
        >>> entry = HistoryEntry(
        ...     timestamp=datetime.now(),
        ...     from_state=State("Off"),
        ...     to_state=State("On"),
        ...     event=Event("TurnOn"),
        ...     success=True
        ... )
        >>> manager.push(entry)
        >>> manager.last()
        HistoryEntry(...)
    """

    def __init__(self, max_entries: int = 1000) -> None:
        """
        Initialize the history manager.

        Args:
            max_entries: Maximum number of entries to keep
        """
        self._max_entries = max_entries
        self._entries: Deque[HistoryEntry] = deque(maxlen=max_entries)
        self._last_state_change: Optional[datetime] = None

    @property
    def max_entries(self) -> int:
        """Get the maximum number of entries."""
        return self._max_entries

    def push(self, entry: HistoryEntry) -> None:
        """
        Add an entry to the history.

        Args:
            entry: The history entry to add
        """
        # Calculate duration if we have a previous entry
        if self._entries and self._last_state_change:
            last_entry = self._entries[-1]
            if last_entry.success:
                duration = (entry.timestamp - self._last_state_change).total_seconds()
                last_entry.duration = duration

        self._entries.append(entry)
        if entry.success:
            self._last_state_change = entry.timestamp

    def last(self) -> Optional[HistoryEntry]:
        """
        Get the most recent entry.

        Returns:
            The most recent entry, or None if empty
        """
        if self._entries:
            return self._entries[-1]
        return None

    def first(self) -> Optional[HistoryEntry]:
        """
        Get the oldest entry.

        Returns:
            The oldest entry, or None if empty
        """
        if self._entries:
            return self._entries[0]
        return None

    def len(self) -> int:
        """
        Get the number of entries.

        Returns:
            Number of entries in history
        """
        return len(self._entries)

    def is_empty(self) -> bool:
        """
        Check if history is empty.

        Returns:
            True if no entries exist
        """
        return len(self._entries) == 0

    def entries(self) -> List[HistoryEntry]:
        """
        Get all entries in chronological order.

        Returns:
            List of history entries
        """
        return list(self._entries)

    def entries_reversed(self) -> List[HistoryEntry]:
        """
        Get all entries in reverse chronological order.

        Returns:
            List of history entries, newest first
        """
        return list(reversed(self._entries))

    def entries_for_state(self, state: State) -> List[HistoryEntry]:
        """
        Get entries involving a specific state.

        Args:
            state: The state to filter by

        Returns:
            List of entries where the state was involved
        """
        return [
            entry for entry in self._entries
            if entry.from_state == state or entry.to_state == state
        ]

    def entries_for_event(self, event: Event) -> List[HistoryEntry]:
        """
        Get entries triggered by a specific event.

        Args:
            event: The event to filter by

        Returns:
            List of entries triggered by the event
        """
        return [
            entry for entry in self._entries
            if entry.event == event
        ]

    def successful_entries(self) -> List[HistoryEntry]:
        """
        Get only successful entries.

        Returns:
            List of successful entries
        """
        return [entry for entry in self._entries if entry.success]

    def failed_entries(self) -> List[HistoryEntry]:
        """
        Get only failed entries.

        Returns:
            List of failed entries
        """
        return [entry for entry in self._entries if not entry.success]

    def clear(self) -> None:
        """Clear all history entries."""
        self._entries.clear()
        self._last_state_change = None

    def format_all(self) -> str:
        """
        Format all entries as a string.

        Returns:
            Formatted string of all entries
        """
        if not self._entries:
            return "No history entries"
        return "\n".join(str(entry) for entry in self._entries)

    def format_summary(self) -> str:
        """
        Format a summary of the history.

        Returns:
            Summary string with statistics
        """
        total = len(self._entries)
        successful = len(self.successful_entries())
        failed = len(self.failed_entries())

        lines = [
            f"History Summary:",
            f"  Total transitions: {total}",
            f"  Successful: {successful}",
            f"  Failed: {failed}",
        ]

        if self._entries:
            first = self.first()
            last = self.last()
            if first and last:
                lines.append(f"  First: {first.timestamp.isoformat()}")
                lines.append(f"  Last: {last.timestamp.isoformat()}")

        return "\n".join(lines)

    def __len__(self) -> int:
        return len(self._entries)

    def __bool__(self) -> bool:
        """Always return True - the manager exists even if empty."""
        return True

    def __repr__(self) -> str:
        return f"HistoryManager(entries={len(self._entries)}, max={self._max_entries})"

File: projects/state-machine/src/history.rs
use crate::{Event, State};
use std::collections::VecDeque;
use std::fmt::Debug;
use std::time::{Duration, Instant};

/// Represents a single history entry
#[derive(Debug, Clone)]
pub struct HistoryEntry<S: State, E: Event> {
    /// Source state before transition
    pub from: S,
    /// Target state after transition
    pub to: S,
    /// Event that triggered the transition
    pub event: E,
    /// Timestamp when the transition occurred
    pub timestamp: Instant,
    /// Duration of the transition (if measured)
    pub duration: Option<Duration>,
    /// Whether the transition was successful
    pub success: bool,
    /// Optional error message if transition failed
    pub error: Option<String>,
}

impl<S: State, E: Event> HistoryEntry<S, E> {
    /// Create a successful history entry
    pub fn success(from: S, to: S, event: E) -> Self {
        Self {
            from,
            to,
            event,
            timestamp: Instant::now(),
            duration: None,
            success: true,
            error: None,
        }
    }

    /// Create a failed history entry
    pub fn failure(from: S, to: S, event: E, error: String) -> Self {
        Self {
            from,
            to,
            event,
            timestamp: Instant::now(),
            duration: None,
            success: false,
            error: Some(error),
        }
    }

    /// Set the duration for this entry
    pub fn with_duration(mut self, duration: Duration) -> Self {
        self.duration = Some(duration);
        self
    }

    /// Get a formatted string representation
    pub fn format(&self) -> String {
        let status = if self.success { "SUCCESS" } else { "FAILURE" };
        let duration_str = self.duration
            .map(|d| format!(" ({:?})", d))
            .unwrap_or_default();

        format!(
            "[{:?}] {:?} --({:?})--> {:?} {}{}",
            self.timestamp, self.from, self.event, self.to, status, duration_str
        )
    }
}

/// Manages history entries for a state machine
pub struct HistoryManager<S: State, E: Event> {
    /// Maximum number of history entries to keep
    max_entries: usize,
    /// History entries
    entries: VecDeque<HistoryEntry<S, E>>,
}

impl<S: State, E: Event> HistoryManager<S, E> {
    /// Create a new history manager with a maximum number of entries
    pub fn new(max_entries: usize) -> Self {
        Self {
            max_entries,
            entries: VecDeque::with_capacity(max_entries),
        }
    }

    /// Add a new history entry
    pub fn push(&mut self, entry: HistoryEntry<S, E>) {
        if self.entries.len() >= self.max_entries {
            self.entries.pop_front();
        }
        self.entries.push_back(entry);
    }

    /// Get the most recent entry
    pub fn last(&self) -> Option<&HistoryEntry<S, E>> {
        self.entries.back()
    }

    /// Get the number of entries
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    /// Check if history is empty
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// Get all entries
    pub fn entries(&self) -> &VecDeque<HistoryEntry<S, E>> {
        &self.entries
    }

    /// Get entries in reverse order (newest first)
    pub fn entries_reversed(&self) -> Vec<&HistoryEntry<S, E>> {
        self.entries.iter().rev().collect()
    }

    /// Get entries for a specific state
    pub fn entries_for_state(&self, state: &S) -> Vec<&HistoryEntry<S, E>> {
        self.entries
            .iter()
            .filter(|entry| entry.from == *state || entry.to == *state)
            .collect()
    }

    /// Get entries for a specific event
    pub fn entries_for_event(&self, event: &E) -> Vec<&HistoryEntry<S, E>> {
        self.entries
            .iter()
            .filter(|entry| entry.event == *event)
            .collect()
    }

    /// Get successful entries only
    pub fn successful_entries(&self) -> Vec<&HistoryEntry<S, E>> {
        self.entries
            .iter()
            .filter(|entry| entry.success)
            .collect()
    }

    /// Get failed entries only
    pub fn failed_entries(&self) -> Vec<&HistoryEntry<S, E>> {
        self.entries
            .iter()
            .filter(|entry| !entry.success)
            .collect()
    }

    /// Clear all history entries
    pub fn clear(&mut self) {
        self.entries.clear();
    }

    /// Get the maximum number of entries
    pub fn max_entries(&self) -> usize {
        self.max_entries
    }

    /// Format all entries as a string
    pub fn format_all(&self) -> String {
        self.entries
            .iter()
            .map(|entry| entry.format())
            .collect::<Vec<_>>()
            .join("\n")
    }
}

impl<S: State, E: Event> Default for HistoryManager<S, E> {
    fn default() -> Self {
        Self::new(100)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
    enum TestState {
        A,
        B,
        C,
    }

    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
    enum TestEvent {
        GoToB,
        GoToC,
        GoToA,
    }

    impl State for TestState {}
    impl Event for TestEvent {}

    #[test]
    fn test_history_entry_success() {
        let entry = HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB);
        assert_eq!(entry.from, TestState::A);
        assert_eq!(entry.to, TestState::B);
        assert_eq!(entry.event, TestEvent::GoToB);
        assert!(entry.success);
        assert!(entry.error.is_none());
    }

    #[test]
    fn test_history_entry_failure() {
        let entry = HistoryEntry::failure(
            TestState::A,
            TestState::B,
            TestEvent::GoToB,
            "guard rejected".to_string(),
        );
        assert!(!entry.success);
        assert_eq!(entry.error, Some("guard rejected".to_string()));
    }

    #[test]
    fn test_history_entry_with_duration() {
        let entry = HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB)
            .with_duration(Duration::from_millis(100));
        assert!(entry.duration.is_some());
        assert_eq!(entry.duration.unwrap(), Duration::from_millis(100));
    }

    #[test]
    fn test_history_entry_format() {
        let entry = HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB);
        let formatted = entry.format();
        assert!(formatted.contains("SUCCESS"));
        assert!(formatted.contains("A"));
        assert!(formatted.contains("B"));
        assert!(formatted.contains("GoToB"));
    }

    #[test]
    fn test_history_manager_new() {
        let manager: HistoryManager<TestState, TestEvent> = HistoryManager::new(50);
        assert_eq!(manager.max_entries(), 50);
        assert!(manager.is_empty());
        assert_eq!(manager.len(), 0);
    }

    #[test]
    fn test_history_manager_push() {
        let mut manager = HistoryManager::new(10);
        let entry = HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB);
        manager.push(entry);
        assert_eq!(manager.len(), 1);
        assert!(!manager.is_empty());
    }

    #[test]
    fn test_history_manager_max_entries() {
        let mut manager = HistoryManager::new(3);
        for _ in 0..5 {
            let entry = HistoryEntry::success(
                TestState::A,
                TestState::B,
                TestEvent::GoToB,
            );
            manager.push(entry);
        }
        assert_eq!(manager.len(), 3);
    }

    #[test]
    fn test_history_manager_last() {
        let mut manager = HistoryManager::new(10);
        assert!(manager.last().is_none());

        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        assert!(manager.last().is_some());
        assert_eq!(manager.last().unwrap().to, TestState::B);
    }

    #[test]
    fn test_history_manager_entries() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::success(TestState::B, TestState::C, TestEvent::GoToC));

        let entries = manager.entries();
        assert_eq!(entries.len(), 2);
    }

    #[test]
    fn test_history_manager_entries_reversed() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::success(TestState::B, TestState::C, TestEvent::GoToC));

        let reversed = manager.entries_reversed();
        assert_eq!(reversed.len(), 2);
        assert_eq!(reversed[0].to, TestState::C);
        assert_eq!(reversed[1].to, TestState::B);
    }

    #[test]
    fn test_history_manager_entries_for_state() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::success(TestState::B, TestState::C, TestEvent::GoToC));
        manager.push(HistoryEntry::success(TestState::C, TestState::A, TestEvent::GoToA));

        let entries_b = manager.entries_for_state(&TestState::B);
        assert_eq!(entries_b.len(), 2);
    }

    #[test]
    fn test_history_manager_entries_for_event() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::success(TestState::B, TestState::C, TestEvent::GoToC));
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));

        let entries_go_to_b = manager.entries_for_event(&TestEvent::GoToB);
        assert_eq!(entries_go_to_b.len(), 2);
    }

    #[test]
    fn test_history_manager_successful_entries() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::failure(
            TestState::B,
            TestState::C,
            TestEvent::GoToC,
            "error".to_string(),
        ));
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));

        assert_eq!(manager.successful_entries().len(), 2);
    }

    #[test]
    fn test_history_manager_failed_entries() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::failure(
            TestState::B,
            TestState::C,
            TestEvent::GoToC,
            "error".to_string(),
        ));
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));

        assert_eq!(manager.failed_entries().len(), 1);
    }

    #[test]
    fn test_history_manager_clear() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::success(TestState::B, TestState::C, TestEvent::GoToC));
        assert_eq!(manager.len(), 2);

        manager.clear();
        assert_eq!(manager.len(), 0);
        assert!(manager.is_empty());
    }

    #[test]
    fn test_history_manager_format_all() {
        let mut manager = HistoryManager::new(10);
        manager.push(HistoryEntry::success(TestState::A, TestState::B, TestEvent::GoToB));
        manager.push(HistoryEntry::success(TestState::B, TestState::C, TestEvent::GoToC));

        let formatted = manager.format_all();
        assert!(formatted.contains("SUCCESS"));
        assert!(formatted.contains("A"));
        assert!(formatted.contains("B"));
        assert!(formatted.contains("C"));
    }

    #[test]
    fn test_history_manager_default() {
        let manager: HistoryManager<TestState, TestEvent> = HistoryManager::default();
        assert_eq!(manager.max_entries(), 100);
    }
}
